cpp/include/cudf/detail/copy_if_else.cuh

```diff
@@ -64,17 +64,14 @@ __launch_bounds__(block_size) __global__
   size_type warp_cur = warp_begin + warp_id;
   size_type index    = tid;
   while (warp_cur <= warp_end) {
-    bool valid = false;
-    if (index >= begin && index < end) {
-      auto value = filter(index) ? lhs[index] : rhs[index];
-      valid      = !has_validity || value.has_value();
-      if (valid) { out.element<T>(index) = static_cast<T>(value.value()); }
-    }
+    auto const opt_value =
+      (index < end) ? (filter(index) ? lhs[index] : rhs[index]) : thrust::nullopt;
+    if (not has_nulls or opt_value) { out.element<T>(index) = static_cast<T>(opt_value.value()); }
```

Comment:
I'm leery of using the `thrust::optional::value` member because it can throw exceptions. However, because it's a `constexpr` function and we're using `--expt-relaxed-constexpr`, nvcc will just discard the exception logic since it's UB. This can lead to surprising behavior. This is precisely what led me down the path of opening this issue: https://github.com/rapidsai/cudf/issues/7795

Long story short, use the `operator*` to access the value without any error checking and potential to throw exceptions.